### PR TITLE
swarm: Refactor and improve HTTP API

### PR DESCRIPTION
**Pull request by lmars**
This PR ~~removes~~ deprecates the file related RPC calls in favour of an improved HTTP API.

The main aim is to expose a simple to use API which can be consumed by thin clients (e.g. `curl` and HTML forms) without the need for complex logic (e.g. manipulating prefix trie manifests).

Summary of changes:

* support uploading multiple files as either a tar stream or multipart form
* support downloading multiple files as a tar stream
* support listing files in a manifest, either as JSON or a browsable HTML document
* introduce `swarm/api/client` which has methods for uploading / downloading single or multiple files and listing files in a manifest
* refactor `swarm/api/http` to be more readable and introduce handling of both tar streams and multipart forms
* add `swarm ls` to list files in a manifest (replaces #3742, ref #3540)

Here are some examples which hopefully show the improvements:

https://gist.github.com/lmars/a37f3eaa129f95273c8c536e98920368

I know this is a big diff, but I found it hard to do this incrementally.

### CODE REVIEW

File: swarm/api/storage.go [hunk 1]
```diff
@@ -1,70 +0,0 @@
-// Copyright 2016 The go-ethereum Authors
```
Comment (lmars) [inline swarm/api/storage.go]: Given swarm is pre 1.0 (I believe 0.2?) and this IPC API is pretty much undocumented, it is totally acceptable to make breaking changes or remove it entirely (this is [encouraged by semver](http://semver.org/#doesnt-this-discourage-rapid-development-and-fast-iteration)). IPC clients can easily switch to the improved HTTP API, and library users can just call `Get` and `Put` on the `api.API` object instead.

At this stage we should really not be trying to incur technical debt, rather iterating until we reach an API we consider stable enough to call 1.0.

File: swarm/api/manifest.go [hunk 2]
```diff
@@ -19,34 +19,167 @@ package api
...
+	ManifestType = "application/bzz-manifest+json"
 )
 
+// Manifest represents a swarm manifest
+type Manifest struct {
+	Entries []ManifestEntry `json:"entries,omitempty"`
+}
+
+// ManifestEntry represents an entry in a swarm manifest
+type ManifestEntry struct {
```
Comment (lmars) [inline swarm/api/manifest.go]: > This looks like a breaking change. If you dont have Path here old manifests will not parse with this?

`ManifestEntry` is embedding `FileInfo` which means it inherits all the fields, so `ManifestEntry.Path` still exists here and will serialize to JSON in the same way, so it will not break old manifests.

> We must put the path outside FileInfo anyway. Path is not a file info in the manifest context, it is a routing instruction ... Same goes for content_type.

What is wrong with having `Path` on `FileInfo`? This makes it very easy for clients to add files with the attributes they want, for example:

* add a file which I can access using the path `some/path`, content type `text/plain` and permissions `0644`:

```go
client.Upload(&FileInfo{
	Path: "some/path",
	ContentType: "text/plain",
	Mode: 0644,
})
```

If this is confusing (perhaps you are thinking of `os.FileInfo`) then perhaps we should choose a different name? Perhaps just drop `FileInfo` and use `ManifestEntry` directly in the clients?

File: swarm/api/api.go [hunk 3]
```diff
@@ -70,86 +71,50 @@ func (self *Api) Store(data io.Reader, size int64, wg *sync.WaitGroup) (key stor
...
-	if err != nil {
-		return nil, "", 500, fmt.Errorf("can't resolve: %v", err)
-	}
-
-	quitC := make(chan bool)
-	trie, err := loadManifest(self.dpa, key, quitC)
+func (self *Api) Get(key storage.Key, path string) (reader storage.LazySectionReader, mimeType string, status int, err error) {
+	trie, err := loadManifest(self.dpa, key, nil)
 	if err != nil {
 		log.Warn(fmt.Sprintf("loadManifestTrie error: %v", err))
```
Comment (lmars) [inline swarm/api/api.go]: Yes I think we need to do a pass over the whole codebase and improve errors and how they are logged, but let's leave that for another PR :wink:

File: swarm/api/manifest.go (hunk 2 above)
Comment (zelig) [inline swarm/api/manifest.go]: I would vote for that yes.
although i feel uncomfortable having Mode directly on a manifest entry, but this is still simplest

**Comment by lmars** [comment]
@zelig I've removed `FileInfo` and just added fields to `ManifestEntry` as well as just deprecating the IPC APIs rather than removing them.

This is ready for another review.

**Comment by fjl** [comment]
Sorry, this is now broken because other stuff was merged. Please rebase.

**Comment by lmars** [comment]
@fjl rebase done.
